### PR TITLE
Omit stack traces when invalid command line flags are passed

### PR DESCRIPTION
**Pull request by manivaradarajan**
* Motivation for features / changes

TensorBoard currently shows stack trace on incorrect command line flag invocation.
Addresses part of #1798.

* Technical description of changes

Wrap this exception with a FlagsException and catch it so basic usage, even if
incorrect, is easier on the eyes.

* Screenshots of UI changes

Captured shell output:

BEFORE:

```
$ bazel-bin/tensorboard/tensorboard
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_maniv/eea943d48c34580f7adce2a516b87755/execroot/org_tensorflow_tensorboard/bazel-out/darwin-fastbuild/bin/tensorboard/tensorboard.runfiles/org_tensorflow_tensorboard/tensorboard/main.py", line 71, in <module>
    run_main()
  File "/private/var/tmp/_bazel_maniv/eea943d48c34580f7adce2a516b87755/execroot/org_tensorflow_tensorboard/bazel-out/darwin-fastbuild/bin/tensorboard/tensorboard.runfiles/org_tensorflow_tensorboard/tensorboard/main.py", line 62, in run_main
    app.run(tensorboard.main, flags_parser=tensorboard.configure)
  File "/Users/maniv/python/tf2.0/lib/python2.7/site-packages/absl/app.py", line 294, in run
    flags_parser,
  File "/Users/maniv/python/tf2.0/lib/python2.7/site-packages/absl/app.py", line 351, in _run_init
    flags_parser=flags_parser,
  File "/Users/maniv/python/tf2.0/lib/python2.7/site-packages/absl/app.py", line 213, in _register_and_parse_flags_with_usage
    args_to_main = flags_parser(original_argv)
  File "/private/var/tmp/_bazel_maniv/eea943d48c34580f7adce2a516b87755/execroot/org_tensorflow_tensorboard/bazel-out/darwin-fastbuild/bin/tensorboard/tensorboard.runfiles/org_tensorflow_tensorboard/tensorboard/program.py", line 194, in configure
    loader.fix_flags(flags)
  File "/private/var/tmp/_bazel_maniv/eea943d48c34580f7adce2a516b87755/execroot/org_tensorflow_tensorboard/bazel-out/darwin-fastbuild/bin/tensorboard/tensorboard.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/core/core_plugin.py", line 461, in fix_flags
    raise ValueError('A logdir or db must be specified. '
ValueError: A logdir or db must be specified. For example `tensorboard --logdir mylogdir` or `tensorboard --db sqlite:~/.tensorboard.db`. Run `tensorboard --helpfull` for details and examples.
```

AFTER:

```
$ bazel-bin/tensorboard/tensorboard
Error: A logdir or db must be specified.

For example "tensorboard --logdir mylogdir"
or `tensorboard --db sqlite:~/.tensorboard.db`.

Run `tensorboard --helpshort` for a list of all options.
Run `tensorboard --helpfull` for additional details and examples.
```

* Detailed steps to verify changes work correctly (as executed by you)

Invoke tensorboard with no arguments and ensure that no stack trace shows up.

* Alternate designs / implementations considered

N/A

### CODE REVIEW

File: tensorboard/main.py [hunk 1]
```diff
@@ -63,6 +64,9 @@ def run_main():
     raise AssertionError("absl.app.run() shouldn't return")
   except ImportError:
     pass
+  except base_plugin.FlagsException as e:
+    print("Error: " + str(e))
```
Comment (wchargin) [inline tensorboard/main.py]: Should probably be printed to stderr? Consider

```py
from tensorboard.util import tb_logging

# ...
except base_plugin.FlagsException as e:
  logger = tb_logging.getLogger()
  logger.error("%s", e)
```

File: tensorboard/main.py [hunk 2]
```diff
@@ -63,6 +64,9 @@ def run_main():
...
+    sys.exit(-1)
```
Comment (wchargin) [inline tensorboard/main.py]: Exiting with `-1` is misleading; the exit status is a u8, so exiting
with `-1` actually yields a WEXITSTATUS of 255. Why not `sys.exit(1)`?

File: tensorboard/plugins/base_plugin.py [hunk 3]
```diff
@@ -196,3 +196,10 @@ def __init__(self, plugin_class):
 
   def load(self, context):
     return self._plugin_class(context)
+
+
+class FlagsException(ValueError):
+  """Raised when a command line flag is not specified or is invalid."""
+  pass
+
```
Comment (wchargin) [inline tensorboard/plugins/base_plugin.py]: nit: remove two trailing newlines

File: tensorboard/plugins/base_plugin.py [hunk 4]
```diff
@@ -196,3 +196,10 @@ def __init__(self, plugin_class):
 
   def load(self, context):
     return self._plugin_class(context)
+
+
+class FlagsException(ValueError):
```
Comment (wchargin) [inline tensorboard/plugins/base_plugin.py]: Please modify the docs for `TBLoader.fix_flags` to indicate that
implementors may raise this exception to signal an error.

File: tensorboard/plugins/core/core_plugin.py [hunk 5]
```diff
@@ -451,17 +450,19 @@ def define_flags(self, parser):
...
             'Must specify either --logdir or --event_file, but not both.')
       if not (flags.logdir or flags.event_file):
-        raise ValueError('Must specify either --logdir or --event_file.')
+        raise FlagsException('Must specify either --logdir or --event_file.')
     elif not flags.db and not flags.logdir:
-      raise ValueError('A logdir or db must be specified. '
-                       'For example `tensorboard --logdir mylogdir` '
-                       'or `tensorboard --db sqlite:~/.tensorboard.db`. '
-                       'Run `tensorboard --helpfull` for details and examples.')
+      raise FlagsException('A logdir or db must be specified.\n\n'
```
Comment (wchargin) [inline tensorboard/plugins/core/core_plugin.py]: I’d prefer to omit the change to this plugin’s particular error message
from this PR, which is otherwise a straightforward refactoring. Are you
open to moving this message change into a separate PR?

File: tensorboard/plugins/core/core_plugin.py [hunk 6]
```diff
@@ -451,17 +450,19 @@ def define_flags(self, parser):
...
+                       'For example "tensorboard --logdir mylogdir"\n'
+                       'or `tensorboard --db sqlite:~/.tensorboard.db`.\n\n'
+                       'Run `tensorboard --helpshort` for a list of all options.\n'
+                       'Run `tensorboard --helpfull` for additional details and examples.\n')
```
Comment (wchargin) [inline tensorboard/plugins/core/core_plugin.py]: nit: These hanging indents have improper indentation. (Before this
change, they were all lined up at the opening parenthesis; after this
change, they’re just straggling.) Either indent them by four more spaces
to match, or change them to non-hanging indents:

```py
raise FlagsException(
    'A logdir or db must be specified.\n\n'
    'For example...'
)
```

I prefer the latter because it’s not brittle to changes that (like this
one) would require the whole block to be reflowed due to a change in
just one line, but the style guide permits either.

Comment (wchargin) [inline tensorboard/plugins/core/core_plugin.py]: Superfluous `\n` on the last line?

File: tensorboard/plugins/base_plugin.py (hunk 4 above)
Comment (wchargin) [inline tensorboard/plugins/base_plugin.py]: User-defined errors in Python have names ending in `Error`, not
`Exception` (just like the built-in exceptions). Rename to `FlagsError`?

File: tensorboard/main.py [hunk 7]
```diff
@@ -48,6 +48,7 @@
 
 from tensorboard import default
 from tensorboard import program
+from tensorboard.plugins import base_plugin
```
Comment (wchargin) [inline tensorboard/main.py]: Please add the corresponding Bazel dependency: `//tensorboard:main`
should depend on `//tensorboard/plugins:base_plugin`.

File: tensorboard/plugins/core/core_plugin.py [hunk 8]
```diff
@@ -23,7 +23,6 @@
 import gzip
 import math
 import mimetypes
-import os
```
Comment (wchargin) [inline tensorboard/plugins/core/core_plugin.py]: Personally I’d put this in a separate PR as well, but I could be
convinced to let this one slide. :-)

File: tensorboard/main.py (hunk 1 above)
Comment (manivaradarajan) [reply tensorboard/main.py]: tb.logging prefixes the message with some goobledygook:

`E0208 21:29:29.716873 140735652410240 main.py:70]`

I've changed to using stderr with print.

File: tensorboard/main.py (hunk 2 above)
Comment (manivaradarajan) [reply tensorboard/main.py]: Changed. This was an old habit! In fact I didn't know that it was a u8.

File: tensorboard/plugins/base_plugin.py (hunk 4 above)
Comment (manivaradarajan) [reply tensorboard/plugins/base_plugin.py]: Added a "Raises: " section to the doc.

Comment (manivaradarajan) [reply tensorboard/plugins/base_plugin.py]: Done

File: tensorboard/plugins/core/core_plugin.py (hunk 5 above)
Comment (manivaradarajan) [reply tensorboard/plugins/core/core_plugin.py]: Sure.

File: tensorboard/plugins/core/core_plugin.py (hunk 6 above)
Comment (manivaradarajan) [reply tensorboard/plugins/core/core_plugin.py]: Moving to a different PR.

Comment (manivaradarajan) [reply tensorboard/plugins/core/core_plugin.py]: Moving to different PR.

File: tensorboard/plugins/core/core_plugin.py (hunk 8 above)
Comment (manivaradarajan) [reply tensorboard/plugins/core/core_plugin.py]: I'll remove it for cleanliness

**Comment by manivaradarajan** [comment]
PTAL

**Comment by wchargin** [comment]
@manivaradarajan: Friendly reminder that this never got merged. :-)
